draft-ietf-jsonpath-base.md

```diff
@@ -1314,21 +1322,94 @@ LCALPHA                 = %x61-7A  ; "a".."z"
 
 function-expression     = function-name "(" S [function-argument
                              *(S "," S function-argument)] S ")"
-function-argument       = comparable / filter-path
+function-argument       = filter-path / comparable
 ~~~
 
-Syntactically, a function-expression can occur anywhere where a
-singular-path can occur (exist-expr, comparable).
-A function-expression that employs a function extension that returns a
-nodelist MUST return a singular node to be used in a comparable.
+A function argument is a `filter-path` or a `comparable`.
+
+According to {{filter-selector}}, a `function-expression` is valid as a `filter-path`
+or a `comparable`.
+
+Any function expressions in a query must be well-formed (by conforming to the above ABNF)
+and correctly typed,
+otherwise the JSONPath implementation MUST raise an error
+(see {{synsem-overview}}).
+To define which function expressions are correctly typed,
+a type system is first introduced.
+
+### Type System for Function Expressions
+
+Each argument and result of a function extension must have a declared type.
+
+A type is a set of instances. A type is a subtype of another type if its set of instances (possibly after coercion)
+is a subset of the set of instances of the other type.
+
+{{tbl-types}} defines the available types in terms of abstract instances, where `n` denotes a node, `v` denotes a value, and `nl` denotes
+a non-empty nodelist. The table also lists the subtypes of each type.
+
+| Type | Abstract Instances | Subtypes |
+| :--: | :----------------: | :------: |
+| `OptionalNodeOrValue` | `Node(n)`, `Value(v)`, `None` | `OptionalNode`, `OptionalValue`, `Value`, `Absent` |
+| `OptionalNode` | `Node(n)`, `None` | `Absent` |
+| `OptionalValue` | `Value(v)`, `None` | `Value`, `Absent` |
```

Comment:
None is used in Python for the JSON value `null`.
We probably want to use a different name so Python users don't think these are the same.

Reply:
Changed to Nothing.